Prevent registering static FASTTLS with multiple libjulias

topolarity opened the pull request:
When there are multiple copies of libjulia present in the application, they can both end up trying to use the same static fast tls storage.

The alternative change here would be to add `jl_options.use_static_fasttls` and modify PackageCompiler.jl to set the flag to `false`.

However, that requires injecting a `#define` in PackageCompiler depending on the presence of the new flag, or starting to assign version numbers to `jl_options_t`

topolarity committed:
Prevent registering static FASTTLS with multiple libjulias

When there are multiple copies of libjulia present in the application,
they can both end up trying to use the same static fast tls storage.

The alternative options I considered for this were:
  - Add `jl_options.use_static_fasttls` and update PackageCompiler.jl to
    set the flag to FALSE
  - Use `dlvsym` in the CLI loader and attach to the exported
    jl_pgcstack_static symbols `__attribute__((symver(...)))`

The second approach is not available on musl, but the first one seems
like the ideal way to handle this long-term.

gbaraldi commented:
I wonder how can we test this? Maybe a small dylib that just tries to access the TLS?

vchuravy commented:
There is also a bit of a design challenge here.

If we have two libjulia's loaded who owns the TLS? It presumably is the outer one. But if the linjulia is of an older (or newer) version it may make assumptions about the data layout behind the TLS. Those assumptions are baked into the generated code.

If we load two libjulia's are we running two GCs? 

I suspect we need to limit it to the same version of Julia (if we expect to run any code within it) or indeed setup a separate TLS (but then our fast TLS for generated code is gonna bit a bit iffy)

topolarity commented:
> If we have two libjulia's loaded who owns the TLS? It presumably is the outer one. But if the linjulia is of an older (or newer) version it may make assumptions about the data layout behind the TLS. Those assumptions are baked into the generated code.

This PR makes it so that the first one loaded will get the fast local-exec TLS, and the other will use its embedding fallback TLS (tied to a static local in each copy of libjulia)

> If we load two libjulia's are we running two GCs?

We are - they are intended to be as isolated as possible for now, with only C-style objects passing between them

vtjnash commented on cli/loader_lib.c:
```diff
@@ -527,8 +527,13 @@ __attribute__((constructor)) void jl_load_libjulia_internal(void) {
     void *(*key)(void) = lookup_symbol(RTLD_DEFAULT, "jl_pgcstack_addr_static");
-    if (fptr != NULL && key != NULL)
-        jl_pgcstack_setkey(fptr, key);
+    _Atomic char *semaphore = lookup_symbol(RTLD_DEFAULT, "jl_pgcstack_static_semaphore");
```
```suggestion
    _Atomic(char) *semaphore = lookup_symbol(RTLD_DEFAULT, "jl_pgcstack_static_semaphore");
```

(apparently on Windows, the `_Atomic x` syntax means something slightly different and is not correctly implemented)

topolarity committed:
Update cli/loader_lib.c

From Jameson:
> apparently on Windows, the _Atomic x syntax means something slightly different and is not correctly implemented

Co-authored-by: Jameson Nash <vtjnash@gmail.com>
